// @telecom-argentina/op-connector/src/accesor/address/mock/address.mock.js
/**
 * Defining mocks
 */

/**
 * @func registerUserMock
 * @description Create a mock response to use API https://dev.openpass.com.ar/bhubApi/1.0.0/address/reverse-external-geolocation
 * 
 * @param {Object} jsonGoogle It's the response object we get when we ask google for an address.
 *
 * @return {Object} An object representing the addres parsed from BHUB
 */
const responseAddressParsedMock = (jsonGoogle) => ({
    id: "306C3A449D0F0EFD216B",
    formatted: "Rafaela 48,2300 Santa Fe Province, Argentina",
    streetNumber: "48",
    postalCode: "2300",
    municipality: {
        id: "31880AFC4B1528A0AE2F",
        description: "Castellanos Department",
        name: "Castellanos Department",
        code: "294"
    },
    city: {
        id: "68B00000000000150931",
        description: "Rafaela",
        name: "Rafaela",
        code: "2095"
    },
    province: {
        id: "0013483030652DBE7F33",
        description: "Santa Fe",
        name: "Santa Fe",
        code: "121"
    },
    country: {
        id: "000B050C1DCD019B9C0E",
        description: "Argentina",
        name: "Argentina",
        code: "ARG"
    }
});


/**
 * Export Mocks
 */

module.exports = {
    responseAddressParsedMock
}
